apis/nucleus/src/components/listbox/ListBox/useSelectionInteractions.js

```diff
@@ -0,0 +1,109 @@
+import { useEffect, useState, useCallback } from 'react';
+import {
+  getUniques,
+  selectValues,
+  applySelectionsOnPages,
+  fillRange,
+  getSelectedValues,
+} from './listbox-selection-utils';
+
+export default function useSelectionInteractions({ layout, selections, model, pages = [] }) {
+  const [mouseDown, setMouseDown] = useState(false);
+  const [selectedElementNumbers, setSelectedElementNumbers] = useState([]);
+  const [selectingValues, setSelectingValues] = useState(false);
+  const [instantPages, setInstantPages] = useState(pages);
+  const [previouslySelected, setPreviouslySelected] = useState([]);
+
+  const onMouseDown = useCallback(
+    (event) => {
+      if (selectingValues) {
+        return;
+      }
+      const elemNumber = +event.currentTarget.getAttribute('data-n');
+      setSelectedElementNumbers([elemNumber]);
+      setMouseDown(true);
+    },
+    [model]
+  );
+
+  // Keeps track of existing selections so we can avoid toggling values on drag select.
+  useEffect(() => {
+    if (mouseDown) {
+      return;
+    }
+    const alreadySelected = getSelectedValues(pages);
+    setPreviouslySelected(alreadySelected);
+  }, [pages, mouseDown]);
+
+  const onMouseUp = useCallback((event) => {
+    if (!mouseDown || selectingValues) {
+      return;
+    }
+    const elemNumber = +event.currentTarget.getAttribute('data-n');
+
+    const isClick = selectedElementNumbers.length === 1;
+
+    setSelectingValues(true);
+
+    const values = getUniques([...selectedElementNumbers.concat(elemNumber)]);
+    const toBeSelected = isClick ? values : values.filter((n) => !previouslySelected.includes(n));
+
+    selectValues({
+      selections,
+      isSingleSelect: layout.qListObject.qDimensionInfo.qIsOneAndOnlyOne,
+      elemNumbers: toBeSelected,
+    }).then(() => {
+      setSelectingValues(false);
+    });
+  });
+
+  document.addEventListener(
+    'mouseup',
+    useCallback(() => {
+      setMouseDown(false);
+    }, [model])
+  );
```

Comment:
I think this should be in a useEffect.

Reply:
Yes, that makes sense. And also unregister the listener there.